### PR TITLE
Changed DocDb Queries for AlarmsByRule to return totalCount of open/acknowledged alarms

### PR DESCRIPTION
**Pull request by jillcary**
# Type of change? 
- [x] Bug fix
- [ ] New feature
- [x] Enhancement
- [x] Breaking change (breaks backward compatibility)

# Description, Context, Motivation 
image

We have a bug, Time control does not pull the correct data #678, where the AlarmByRule counts are returned in a funny way because we only grab the first 1000 alarm results. 
**Checklist:**

- [x] All tests passed
- [x] The code follows the code style and conventions of this project
- [ ] The change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: Services/Alarms.cs [hunk 1]
```diff
@@ -121,15 +139,15 @@ public Alarm Get(string id)
...
             string[] devices)
         {
-            string sql = QueryBuilder.BuildSql(
-                "alarm",
-                id, "rule.id",
-                from, "created",
-                to, "created",
-                order, "created",
+            string sql = QueryBuilder.GetDocumentsSql(
+                ALARM_SCHEMA_KEY,
```
Comment (timlaverty) [inline Services/Alarms.cs]: nit: I have a hard time seeing 9 parameters passed to a function.  I'd suggest putting these in an struct or object and passing the as a group instead.  This is especially the case when I see the same parameter being passed multiple times (MESSAGE_RECEIVED_KEY).

File: Services/Alarms.cs [hunk 2]
```diff
@@ -191,16 +241,14 @@ private Document GetDocumentById(string id)
                 null,
                 "SELECT * FROM c WHERE c.id='" + id + "'",
```
Comment (timlaverty) [inline Services/Alarms.cs]: I wouldn't suggest using * as the projection in any SQL query.  1) It is better to use an explicit list of fields you're returning, 2) giving it * means that you're returning every column in c; are you sure this is what you want?, 3) it's less readable - the maintainer has to go fish for what the projection is.

File: Services/Alarms.cs [hunk 3]
```diff
@@ -191,16 +241,14 @@ private Document GetDocumentById(string id)
...
                 0,
-                1000);
+                DOC_QUERY_LIMIT);
```
Comment (timlaverty) [inline Services/Alarms.cs]: 👍

File: Services/Helpers/QueryBuilder.cs [hunk 4]
```diff
@@ -2,12 +2,17 @@
...
 
 namespace Microsoft.Azure.IoTSolutions.DeviceTelemetry.Services.Helpers
 {
     public class QueryBuilder
     {
-        public static string BuildSql(
+        private const string INVALID_CHARACTER = @"[^A-Za-z0-9:;.,_\-]";
+
+        public static string GetDocumentsSql(
             string schemaName,
```
Comment (timlaverty) [inline Services/Helpers/QueryBuilder.cs]: glug - pass in object?

Comment (timlaverty) [reply Services/Helpers/QueryBuilder.cs]: Also, if we're not going to pass in object, order them so the operators (skip, limit) are last; i.e. things with semantic meaning are grouped.

File: Services/Helpers/QueryBuilder.cs [hunk 5]
```diff
@@ -39,27 +44,104 @@ public class QueryBuilder
             if (from.HasValue)
             {
                 // TODO: left operand is never null
-                DateTimeOffset fromDate = from ?? default(DateTimeOffset);
+                var fromDate = from ?? default(DateTimeOffset);
```
Comment (timlaverty) [inline Services/Helpers/QueryBuilder.cs]: why did you need to go to var from explicit typing?

File: Services/Helpers/QueryBuilder.cs [hunk 6]
```diff
@@ -39,27 +44,104 @@ public class QueryBuilder
...
+            }
+
+            queryBuilder.Append(")");
+
             return queryBuilder.ToString().Replace("`", "\"");
         }
+
+        private static void ValidateInput(ref string input)
+        {
+            input = input.Trim();
```
Comment (timlaverty) [inline Services/Helpers/QueryBuilder.cs]: this creates a new trimmed string and modifies the string in the pointer passed - is this what you mean to do?  

https://msdn.microsoft.com/en-us/library/t97s7bs3(v=vs.95).aspx

File: Services/Helpers/QueryBuilder.cs [hunk 7]
```diff
@@ -39,27 +44,104 @@ public class QueryBuilder
...
+            string fromProperty,
+            DateTimeOffset? to,
+            string toProperty,
+            string[] devices,
+            string devicesProperty,
+            string[] filterValues,
+            string filterProperty) 
+        {
+            var validateDeviceIds = string.Join(",", devices);
+            var validateFilters = string.Join(",", filterValues);
```
Comment (timlaverty) [inline Services/Helpers/QueryBuilder.cs]: nit: could pass this to ValidateInput as string.Join(",", filterValues) instead of creating a local that you thereafter don't use.

File: Services/Helpers/QueryBuilder.cs [hunk 8]
```diff
@@ -39,27 +44,104 @@ public class QueryBuilder
...
+            }
+
+            if (to.HasValue)
+            {
+                // TODO: left operand is never null
+                var toDate = to ?? default(DateTimeOffset);
+                queryBuilder.Append(" AND c[`" + toProperty + "`] <= " + toDate.ToUnixTimeMilliseconds());
+            }
+
+            if (filterValues.Length > 0)
```
Comment (timlaverty) [inline Services/Helpers/QueryBuilder.cs]: Did you mean to use filterValues or validateFilters here?  ...and did you want the resulting joined string to be trimmed?The latter has been joined, passed your validate function, and also been trimmed.  This array is going to get joined but not trimmed.

File: Services/Rules.cs [hunk 9]
```diff
@@ -154,6 +166,49 @@ public class Rules : IRules
...
+            int limit,
+            string[] devices)
+        {
+            var alarmCountByRuleList = new List<AlarmCountByRule>();
+
+            // get list of rules
+            var rulesList = await this.GetListAsync(order, skip, limit, null);
+
+            // get open alarm count and most recent alarm for each rule
+            foreach (var rule in rulesList)
```
Comment (timlaverty) [inline Services/Rules.cs]: the lack of a docdb groupby function make me cry a little.  Can you add a TODO to revisit this code and also note that we need to discuss what we're writing with the docdb team?

File: Services/Rules.cs [hunk 10]
```diff
@@ -184,5 +239,31 @@ public class Rules : IRules
...
+                to,
+                "desc",
+                0,
+                1,
+                devices);
+
+            if (resultList.Count != 0)
+            {
+                return resultList[0];
+            }
```
Comment (timlaverty) [inline Services/Rules.cs]: nit else {  ...}  ?

File: WebService/v1/Models/RuleApiModel.cs [hunk 11]
```diff
@@ -17,7 +17,7 @@ public class RuleApiModel
         [JsonProperty(PropertyName = "Name")]
         public string Name { get; set; }
 
-        [JsonProperty(PropertyName = "DateCreated")]
+        [JsonProperty(PropertyName = "MessageTime")]
```
Comment (dluc) [inline WebService/v1/Models/RuleApiModel.cs]: this is a breaking change, are you sure the Web UI won't break while looking for a different field name?

File: WebService/v1/Models/AlarmApiModel.cs [hunk 12]
```diff
@@ -19,7 +19,7 @@ public class AlarmApiModel
         [JsonProperty(PropertyName = "Id")]
         public string Id { get; set; }
 
-        [JsonProperty(PropertyName = "DateCreated")]
+        [JsonProperty(PropertyName = "MessageTime")]
```
Comment (dluc) [inline WebService/v1/Models/AlarmApiModel.cs]: will this break the web UI?

File: WebService/v1/Controllers/AlarmsByRuleController.cs [hunk 13]
```diff
@@ -19,18 +20,21 @@ public class AlarmsByRuleController : Controller
...
             ILogger logger)
         {
             this.alarmService = alarmService;
+            this.ruleService = ruleService;
             this.log = logger;
         }
 
         [HttpGet]
-        public AlarmByRuleListApiModel List(
+        public async Task<AlarmByRuleListApiModel> List(
```
Comment (dluc) [inline WebService/v1/Controllers/AlarmsByRuleController.cs]: nit: add `Async` method suffix

File: Services/StorageClient.cs [hunk 14]
```diff
@@ -240,6 +248,36 @@ public DocumentClient GetDocumentClient()
...
+            }
+
+            string collectionLink = string.Format(
+                "/dbs/{0}/colls/{1}",
+                databaseName,
+                colId);
+
+            long count = 0;
+
+            foreach (var result in this.client.CreateDocumentQuery(collectionLink, queryString, queryOptions))
```
Comment (dluc) [inline Services/StorageClient.cs]: I assume there is only one result in the list and thus a loop is not required? otherwise it looks weird that you loop through the entire collection and keep rewriting the value of `count`

File: Services/Rules.cs [hunk 15]
```diff
@@ -184,5 +239,31 @@ public class Rules : IRules
...
+
+            string msg = "Could not retrieve most recent alarm " +
+                         "for rule id " + id;
+            this.log.Debug(msg, () => new { });
```
Comment (dluc) [inline Services/Rules.cs]:
```
this.log.Debug("Could not retrieve most recent alarm", () => new { id });
```

File: Services/Helpers/QueryBuilder.cs [hunk 16]
```diff
@@ -39,27 +44,104 @@ public class QueryBuilder
...
             return queryBuilder.ToString().Replace("`", "\"");
         }
+
+        private static void ValidateInput(ref string input)
+        {
+            input = input.Trim();
+
+            if (Regex.IsMatch(input, INVALID_CHARACTER))
+            {
+                throw new InvalidInputException("input \'" +
```
Comment (dluc) [inline Services/Helpers/QueryBuilder.cs]: why quote the single quote? also, I think the code is more readable in a single line in this case

File: Services/Alarms.cs (hunk 1 above)
Comment (dluc) [reply Services/Alarms.cs]: we have the same problem in the Java implementation btw, I agree this increases the risk of bugs plus makes the code hard to read/review

File: Services/StorageClient.cs [hunk 17]
```diff
@@ -240,6 +248,36 @@ public DocumentClient GetDocumentClient()
             return docs;
         }
 
+        public int QueryCount(
+            string databaseName,
+            string colId,
+            FeedOptions queryOptions,
+            string queryString,
+            int skip,
+            int limit)
```
Comment (dluc) [inline Services/StorageClient.cs]: skip and limit are never used

File: Services/Helpers/QueryBuilder.cs (hunk 5 above)
Comment (jillcary) [reply Services/Helpers/QueryBuilder.cs]: Was following ReSharper suggestion for style.

File: Services/Helpers/QueryBuilder.cs (hunk 6 above)
Comment (jillcary) [reply Services/Helpers/QueryBuilder.cs]: Yes.

File: Services/Helpers/QueryBuilder.cs (hunk 7 above)
Comment (jillcary) [reply Services/Helpers/QueryBuilder.cs]: The ` character is required for the query, but not valid input. 
due to the ref, ValidateInput(validateDeviceIds) is not valid syntax. 
Added issue for a general input validation clean up task to address this all-up for the service. 
https://github.com/Azure/device-telemetry-dotnet/issues/55

File: Services/Helpers/QueryBuilder.cs (hunk 16 above)
Comment (jillcary) [reply Services/Helpers/QueryBuilder.cs]: Updated.

File: Services/Helpers/QueryBuilder.cs (hunk 5 above)
Comment (timlaverty) [reply Services/Helpers/QueryBuilder.cs]: they like var! ok... I like the explicit typing better but your call :).

File: Services/Helpers/QueryBuilder.cs (hunk 4 above)
Comment (jillcary) [reply Services/Helpers/QueryBuilder.cs]: https://github.com/Azure/device-telemetry-dotnet/issues/56

File: Services/Alarms.cs (hunk 1 above)
Comment (jillcary) [reply Services/Alarms.cs]: https://github.com/Azure/device-telemetry-dotnet/issues/56

File: Services/Alarms.cs (hunk 2 above)
Comment (jillcary) [reply Services/Alarms.cs]: https://github.com/Azure/device-telemetry-dotnet/issues/57

File: Services/Helpers/QueryBuilder.cs (hunk 8 above)
Comment (jillcary) [reply Services/Helpers/QueryBuilder.cs]: filter values is the intended check as that's the array that's passed in as a parameter. 
It will get joined but not trimmed, added https://github.com/Azure/device-telemetry-dotnet/issues/55 for general input validation cleanup.

File: Services/Rules.cs (hunk 9 above)
Comment (jillcary) [reply Services/Rules.cs]: See note in QueryBuilder, issue added in comment 
https://github.com/Azure/device-telemetry-dotnet/issues/58

File: Services/Rules.cs (hunk 10 above)
Comment (jillcary) [reply Services/Rules.cs]: ReSharper didn't like my else either :(

File: Services/StorageClient.cs (hunk 17 above)
Comment (jillcary) [reply Services/StorageClient.cs]: Removed.

File: Services/StorageClient.cs (hunk 14 above)
Comment (jillcary) [reply Services/StorageClient.cs]: Updated

File: WebService/v1/Controllers/AlarmsByRuleController.cs (hunk 13 above)
Comment (jillcary) [reply WebService/v1/Controllers/AlarmsByRuleController.cs]: Updated.

File: Services/Rules.cs (hunk 10 above)
Comment (dluc) [reply Services/Rules.cs]: I liked it better like R# suggests :-P
